sympy/solvers/ode/tests/test_riccati.py

```diff
@@ -0,0 +1,903 @@
+    solve_riccati)
+
+f = Function('f')
+x = symbols('x')
+
+# These are the functions used to generate the tests
+# SHOULD NOT BE USED DIRECTLY IN TESTS
+
+def rand_rational(maxint):
+    return Rational(randint(-maxint, maxint), randint(1, maxint))
+
+def rand_poly(x, degree, maxint):
+    return Poly([rand_rational(maxint) for _ in range(degree+1)], x)
+
+def rand_rational_function(x, degree, maxint):
+    degnum = randint(1, degree)
+    degden = randint(1, degree)
+    num = rand_poly(x, degnum, maxint)
+    den = rand_poly(x, degden, maxint)
+    while den == Poly(0, x):
+        den = rand_poly(x, degden, maxint)
+    return num / den
+
+def find_riccati_ode(ratfunc, x, yf):
+    y = ratfunc
+    yp = y.diff(x)
+    q1 = rand_rational_function(x, 1, 3)
+    q2 = rand_rational_function(x, 1, 3)
+    while q2 == 0:
+        q2 = rand_rational_function(x, 1, 3)
+    q0 = ratsimp(yp - q1*y - q2*y**2)
+    eq = Eq(yf.diff(), q0 + q1*yf + q2*yf**2)
+    sol = Eq(yf, y)
+    assert checkodesol(eq, sol) == (True, 0)
+    return eq, q0, q1, q2
+
+
+# Testing functions start
+
+def test_riccati_transformation():
+    tests = [
+    (
+        x/(x - 1),
+        (x**2 + 7)/3*x,
+        x,
+        -x**2/(x - 1) - x*(x**2/3 + S(7)/3)/2 - 1/(2*x)
+    ),
+    (
+        (-2*x**2 - 2*x - 1)/(4*x),
+        (2*x + 1)/(x - 2),
+        -1/(x + 2),
+        Mul(-1, -x/2 - 1, evaluate=False)/(x + 2)**2 - (2*x + 1)/(Mul(2, x - 2, evaluate=False)) \
+        + (-2*x**2 - 2*x - 1)/(4*x*(x + 2))
+    ),
+    (
+        (2*x + 3)/(2*x + 2),
+        (3 - 3*x)/(x + 1),
+        5*x,
+        -5*x*(2*x + 3)/(2*x + 2) - (3 - 3*x)/(Mul(2, x + 1, evaluate=False)) - 1/(2*x)
+    ),
+    (
+        (-2*x - 1)/(2*x**2 + 2*x - 2),
+        -2/x,
+        (-x - 1)/(4*x),
+        -2*x*(-1/(4*x) - (-x - 1)/(4*x**2))/(-x - 1) - (-2*x - 1)*(-x - 1)/(4*x*(2*x**2 + 2*x \
+        - 2)) + 1/x
+    ),
+    (
+        -1/(2*x**2 - 1),
+        0,
+        (2 - x)/(4*x - 2),
+        (2 - x)/((4*x - 2)*(2*x**2 - 1)) - (4*x - 2)*(Mul(-4, 2 - x, evaluate=False)/(4*x - \
+        2)**2 - 1/(4*x - 2))/(Mul(2, 2 - x, evaluate=False))
+    ),
+    (
+        x,
+        (8*x - 12)/(12*x + 9),
+        x**3/(6*x - 9),
+        -x**4/(6*x - 9) - (8*x - 12)/(Mul(2, 12*x + 9, evaluate=False)) - (6*x - 9)*(-6*x**3/(6*x \
+        - 9)**2 + 3*x**2/(6*x - 9))/(2*x**3)
+    ),
+    (
+        -7*x**3 + 4*x + 9,
+        (x - 1)/x,
+        5*x - 1/(x - 1),
+        Mul(-1, 5 + (x - 1)**(-2), evaluate=False)/(10*x - 2/(x - 1)) + (-5*x + 1/(x - 1))*\
+        (-7*x**3 + 4*x + 9) - (x - 1)/(2*x)
+    ),
+    (
+        -6/(x**2 - 2*x + 1),
+        (x + 12)/((x + 1)*(x - 3)),
+        2*x + 5,
+        Mul(-6, -2*x - 5, evaluate=False)/(x**2 - 2*x + 1) - 2/(4*x + 10) - (x + 12)/(\
+        Mul(2, x - 3, evaluate=False)*(x + 1))
+    )]
+    for w, b1, b2, y in tests:
+        assert y == riccati_normal(w, x, b1, b2)
+        assert w == riccati_inverse_normal(y, x, b1, b2).cancel()
```

Comment:
Do we need so many examples for testing these functions? What exactly do the different examples represent?

These functions seem to be relatively straight-forward so I don't see why 8 test cases are needed. A more relevant test would be one that involved the bp parameter for example which doesn't seem to be tested here.

Reply:
The `bp` parameter is used when multiple solutions for the same equation are transformed back in the last step of the algorithm. It is useful only when 2 or more test cases have the same `b1` and `b2`.

Reply:
Yes, but there should be a test that the bp parameter works correctly. If the function has two possible modes of operation then there needs to be a test for each rather than 10 tests for one and none for the other.